src/masternode-budget.cpp

```diff
@@ -1405,23 +1408,38 @@ bool CBudgetProposal::UpdateValid(int nCurrentHeight, bool fCheckCollateral)
         return false;
     }
 
+    int nProposalEnd = nBlockStart + (Params().GetConsensus().nBudgetCycleBlocks + 1) * GetTotalPaymentCount();
+    if (nBlockEnd != nProposalEnd) {
+        strInvalid = "Proposal " + strProposalName + ": Invalid nBlockEnd (mismatch with payments count)";
+        return false;
+    }
+
+    return true;
+}
+
+bool CBudgetProposal::CheckAmount(const CAmount& nTotalBudget)
+{
+    // check minimum amount
     if (nAmount < 10 * COIN) {
-        strInvalid = "Proposal " + strProposalName + ": Invalid nAmount";
+        strInvalid = "Proposal " + strProposalName + ": Invalid nAmount (too low)";
         return false;
     }
 
-    if (address == CScript()) {
-        strInvalid = "Proposal " + strProposalName + ": Invalid Payment Address";
+    // check maximum amount
+    // can only pay out 10% of the possible coins (min value of coins)
+    if (nAmount > nTotalBudget) {
+        strInvalid = "Proposal " + strProposalName + ": Invalid nAmount (too high)";
         return false;
     }
 
-    if (fCheckCollateral) {
-        int nConf = 0;
-        std::string strError;
-        if (!IsBudgetCollateralValid(nFeeTXHash, GetHash(), strError, nTime, nConf)) {
-            strInvalid = "Proposal " + strProposalName + ": Invalid collateral (" + strError + ")";
-            return false;
-        }
+    return true;
+}
+
+bool CBudgetProposal::CheckAddress()
+{
+    if (address == CScript()) {
```

Comment:
would love to see a `isNull` here.

Reply:
updating.

Reply:
This is already changed in the third commit (be2ecf1ca5b202b767ccc5718e0daf2c2b27089e)